### PR TITLE
fix: invalid detection of duplicated relationships during Tracker import

### PR DESCRIPTION
**Pull request by luciano-fiandesio**
This PR introduces a new Tracker Import PreProcessor to remove duplicated relationship from the
Tracker payload (Tracker Bundle). The duplicated relationship logic is detailed here: https://jira.dhis2.org/browse/DHIS2-9965

Additionally, the Relationship validation code (`RelationshipsValidationHook`) has been improved to better report invalid relationships.

- Missing "from" returns `4007` code and message "Missing required relationship property: 'from'."
- Missing "to" returns `4008` code and message "Missing required relationship property: 'to'."
- Missing "relationshipType" returns `4004` code and message "Missing required relationship property: 'relationshipType'."
- Invalid "relationshipType" returns `4009` code and message "Relationship Type `{0}` is not valid." - where 0 corresponds to the relationship type uid
- On invalid Relationship constraint (based on the Relationship Type) the system returns `4010` code and message: ""Relationship Type `{0}` constraint requires a {1} but a {2} was found."
- On multiple value provided for a Relationship Item (e.g. trackerEntity and Enrollment) the system returns `E4001` and message: "Relationship Item `{0}` for Relationship `{1}` is invalid: an Item can link only one Tracker entity." where `{0}` is from|to, `{1}` is the relationship uid

Finally, the constraint and "self-reference" validation only takes place if the required data validation has no errrors.

REF: DHIS2-9965

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java [hunk 1]
```diff
@@ -89,37 +87,57 @@ public void validateRelationship( ValidationErrorReporter reporter, Relationship
...
-            .map( r -> new ImmutablePair<>( r.getTo(), r.getFrom() ) )
-            .collect( Collectors.toList() );
+            } );
+    }
 
-        relationships.addAll( bidirectionalRelationship );
+    private boolean validateMandatoryData( ValidationErrorReporter reporter, Relationship relationship,
+        List<RelationshipType> relationshipsTypes )
+    {
+        addErrorIfNull( relationship.getFrom(), reporter, E4007, "relationship.getFrom" );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java:123]: Here we use "get"+"From"/"To", while further down we just use "relationshipType", maybe we should omit the "get" part? Looks better in my opinion:

relationship.from, relationship.to, relationship.relationshiptype

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java [hunk 2]
```diff
@@ -130,128 +148,58 @@ private void validateAutoRelationship( ValidationErrorReporter reporter, Relatio
...
-            }
+            return EVENT;
         }
-
+        return null;
     }
 
-    private List<TrackerErrorReport.TrackerErrorReportBuilder> validateRelationshipConstraint( RelationshipItem item,
-        RelationshipConstraint relationshipType )
+    private List<TrackerErrorReport.TrackerErrorReportBuilder> validateRelationshipConstraint( String relSide,
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java:168]: This is just a note, more than feedback:

I was thinking we have to validate that wrong values are not inserted as well (IE. tei inserted, when constraint is EVENT).

However, we actually ignore other values when comitting, so I say we don't add that validation, unless you think it makes sense to report sending too much data (wrong data is already covered, assuming wrong data = wrong constraint type)

**Review by stian-sandvold — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerErrorCode.java [hunk 3]
```diff
@@ -145,11 +145,14 @@
     E4000( "Relationship: `{0}` cannot link to itself" ),
     E4001( "Property `{0}` can not be set when property `{1}` is `{2}`. " ),
     E4002( "Property `{0}` must be set when property `{1}` is `{2}`. " ),
-    E4003( "There are duplicated relationships. " ),
-    E4004( "Missing required property: 'relationshipType'." ),
+    E4003( "There are duplicated relationships." ),
+    E4004( "Missing required relationship property: 'relationshipType'." ),
     E4005( "RelationShip: `{0}`, do not exist." ),
     E4006( "Could not find RelationShipType: `{0}`."  ),
```
Comment (vilkg) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerErrorCode.java:151]: Could not find RelationShipType: `{0}`."  -> can you change to `relationshipType`?

Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerErrorCode.java:151]: 🚢

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java (hunk 1 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java:123]: Fixed, the argument was actually ignored. The error message is shown in the PR commit message.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java (hunk 2 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/RelationshipsValidationHook.java:168]: @stian-sandvold Ah, I didn't know. I'll ping you so we can discuss.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerErrorCode.java (hunk 3 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerErrorCode.java:151]: :) gone

**Review by enricocolasante — APPROVED** [review]

**Pull request merged**